README.rst
PoV admin tools
===============

.. image:: https://github.com/ProgrammersOfVilnius/pov-admin-tools/workflows/build/badge.svg?branch=master
    :target: https://github.com/ProgrammersOfVilnius/pov-admin-tools/actions

This package contains a few scripts we use at PoV for managing servers:

- new-changelog-entry_: keeps a sysadmin diary (/root/Changelog)

- check-changelog: checks that timestamps in /root/Changelog are strictly
  increasing

It used to ship a few more scripts, but those got moved to pov-server-page_.

.. _pov-server-page: https://github.com/ProgrammersOfVilnius/pov-server-page

It also Recommends: a bunch of other packages I consider to be indispensible.

Suggested steps for setting up a new server::

    $ sudo -s
    # apt-get update
    # apt-get install software-properties-common
    # add-apt-repository -y ppa:pov
    # apt-get update
    # apt-get install pov-admin-tools
    # apt-get install collectd --no-install-recommends
    # new-changelog-entry

(then copy and paste this list of commands into the changelog)


new-changelog-entry
===================

When run without arguments, appends the current date, time and your username
to ``/root/Changelog`` and launches ``vim`` for you to `describe what you're
doing <https://mg.pov.lt/blog/sysadmin-diary.html>`__.

Alternatively you can pass a short message directly on the command line::

    $ new-changelog-entry "apt-get install apache2"
    [sudo] password for mg:
    2013-07-18 19:09 +0300: mg
      apt-get install apache2

You can append to an existing message by using ``-a``::

    $ new-changelog-entry -a "apt-get install postfix"
    2013-07-18 19:09 +0300: mg
      apt-get install apache2
      apt-get install postfix

You can launch the text editor without appending anyting by using ``-e``.


debian/changelog
pov-admin-tools (0.16.3) UNRELEASED; urgency=medium

  * debian/control:
    - add lsof to Recommends.

 -- Marius Gedminas <marius@gedmin.as>  Sat, 20 Feb 2021 11:39:07 +0200

pov-admin-tools (0.16.2) xenial; urgency=medium

  * check-changelog 1.0.1:
    - port to Python 3,
    - print errors to stderr instead of stdout.
  * debian/control:
    - depend on python3 instead of python,
    - build-depend on python3-pytest,
    - stop recommending ntp, it's unnecessary now that systemd-timesyncd
      exists.
  * debian/rules:
    - override dh_auto_test because tox is unsuitable for the network-isolated
      build environment.

 -- Marius Gedminas <marius@gedmin.as>  Tue, 27 Aug 2019 15:59:34 +0300

pov-admin-tools (0.16.1) xenial; urgency=medium

  * debian/control:
    - drop apt-checkupdates from Recommends: -- uanttended-upgrades supercedes
      it.

 -- Marius Gedminas <marius@gedmin.as>  Thu, 01 Aug 2019 15:40:44 +0300

pov-admin-tools (0.16.0) trusty; urgency=medium

  * Moved machine-summary, disk-inventory and du-diff scripts into
    pov-server-page.
  * debian/control:
    - add Breaks: for old versions of pov-server-page, which expect to find
      du-diff in $PATH.
  * new-changelog-entry 1.2.0:
    - allow leading -- to disambiguate messages that look like filenames.

 -- Marius Gedminas <marius@gedmin.as>  Sat, 20 Jan 2018 18:37:29 +0200

pov-admin-tools (0.15.2) precise; urgency=medium

  * disk-inventory 1.2.2:
    - fix ValueError when there are new physical volumes not assigned to a
      volume group yet.

 -- Marius Gedminas <marius@gedmin.as>  Mon, 13 Mar 2017 20:11:43 +0200

pov-admin-tools (0.15.1) precise; urgency=medium

  * all the scripts now support --version,
  * documented machine-summary -n in the manual page,
  * updated forgotten versions and dates in manual pages,
  * debian/control:
    - pstree is actually in the psmisc package,
    - add command-not-found to recommends.

 -- Marius Gedminas <marius@gedmin.as>  Thu, 20 Oct 2016 10:27:01 +0300

pov-admin-tools (0.15.0) precise; urgency=medium

  * disk-inventory 1.2.1:
    - recognize LVM volumes used for swap.
  * machine-summary 0.7.0:
    - allow -n as short option for --no-title,
    - drop Python 2.6 compatibility,
    - restore Python 3 compatibility,
    - ignore docker* and vnet* (libvirt) virtual network devices,
    - ignore IP addresses on virtual network devices,
    - show which device has which IP address.
  * debian/control:
    - add dstat to recommends.

 -- Marius Gedminas <marius@gedmin.as>  Wed, 21 Sep 2016 08:56:51 +0300

pov-admin-tools (0.14.0) precise; urgency=medium

  * disk-inventory 1.2.0:
    - show RAID device names,
    - handle the root partition being an LVM volume correctly.

 -- Marius Gedminas <marius@gedmin.as>  Wed, 31 Aug 2016 10:37:06 +0300

pov-admin-tools (0.13.2) precise; urgency=medium

  * debian/control:
    - Add pstree and apt-show-versions to recommends.

 -- Marius Gedminas <marius@gedmin.as>  Thu, 04 Feb 2016 18:21:54 +0200

pov-admin-tools (0.13.1) precise; urgency=medium

  * machine-summary 0.6.1:
    - strip netmask from IP address.
  * debian/control:
    - Add pydf to recommends.
    - Add debian-goodies to recommends (for checkrestart).

 -- Marius Gedminas <marius@gedmin.as>  Tue, 08 Dec 2015 09:00:26 +0200

pov-admin-tools (0.13.0) precise; urgency=medium

  * debian/control:
    - Recommend git instead of git-core, vim instead of vim-gnome,
      collectd-core instead of collectd (GH #3).
  * disk-inventory 1.1.8:
    - don't list partition holders, whatever those are ("dm-0 dm-1 dm-2 ..."),
    - align columns when names of LVM volumes or mountpoints are longer than
      expected.

 -- Marius Gedminas <marius@gedmin.as>  Wed, 16 Sep 2015 09:39:21 +0300

pov-admin-tools (0.12.2) precise; urgency=medium

  * disk-inventory 1.1.7:
    - suppress df error on Ubuntu 12.04 LTS.
  * new-changelog-entry 1.1.0:
    - new option (-e) to edit the changelog file.

 -- Marius Gedminas <marius@gedmin.as>  Tue, 23 Jun 2015 10:37:02 +0300

pov-admin-tools (0.12.1) precise; urgency=medium

  * new-changelog-entry 1.0.4:
    - don't forget to add a trailing newline (GH #5).

 -- Marius Gedminas <marius@gedmin.as>  Wed, 22 Apr 2015 11:29:37 +0300

pov-admin-tools (0.12) precise; urgency=medium

  * new-changelog-entry 1.0.3:
    - don't interpret backslash-escapes in the message (GH #4).
  * machine-summary 0.6:
    - new option: --no-title,
    - show IP addresses (GH #1).

 -- Marius Gedminas <marius@gedmin.as>  Tue, 21 Apr 2015 13:18:13 +0300

pov-admin-tools (0.11) precise; urgency=medium

  * Add update-notifier-common to Recommends.
  * machine-summary 0.5.2:
    - handle OpenVZ container virtual disk devices more correctly (GH #2),
    - handle KVM virtual disk devices (GH #2).

 -- Marius Gedminas <marius@gedmin.as>  Fri, 17 Apr 2015 20:05:49 +0300

pov-admin-tools (0.10.1) precise; urgency=medium

  * disk-inventory:
    - include fix for swap device recognition, dropped because I forgot to git
      pull before releasing.

 -- Marius Gedminas <marius@gedmin.as>  Fri, 10 Oct 2014 10:57:51 +0300

pov-admin-tools (0.10) precise; urgency=low

  * machine-summary 0.5.1:
    - handle Xen virtual disk devices correctly
  * disk-inventory 1.1.6:
    - handle Xen virtual disk devices correctly
    - handle HP CCISS RAID devices correctly
  * Add atop and debsums to Recommends.

 -- Marius Gedminas <marius@gedmin.as>  Fri, 10 Oct 2014 10:56:09 +0300

pov-admin-tools (0.9) precise; urgency=low

  * machine-summary 0.5:
    - make it work in Virtuozzo containers
  * Add pov-simple-backup to Recommends.

 -- Marius Gedminas <marius@gedmin.as>  Tue, 15 Oct 2013 11:57:58 +0300

pov-admin-tools (0.8) precise; urgency=low

  * du-diff 0.2:
    - fix error when run without arguments
    - don't strip trailing spaces from file names
  * New script: check-changelog.
  * Add tig and htop to Recommends.


tests.py
import sys
import textwrap

import pytest

import check_changelog as cc


def test_check(capsys):
    cc.check(textwrap.dedent("""\
        2019-08-27
          started this test suite

        2019-08-27 15:35
          wrote some tests

        2019-08-27 15:36
          started using copy paste a lot

        2019-08-27 15:35
          wrote some tests
    """).splitlines(True), 'Changelog')

    out, err = capsys.readouterr()
    assert err == (
        'Changelog:10: date out of order (2019-08-27 15:35'
        ' after 2019-08-27 15:36)\n'
        'Changelog:4: 2019-08-27 15:35 previously used here\n'
    )


def test_main_prints_version(monkeypatch, capsys):
    monkeypatch.setattr(sys, 'argv', ['check-changelog', '--version'])
    with pytest.raises(SystemExit):
        cc.main()
    out, err = capsys.readouterr()
    # Python 2 spits --version to stderr, Python 3 fixes that
    assert out + err == (
        '%s\n' % cc.__version__
    )


def test_main(monkeypatch, capsys):
    monkeypatch.setattr(sys, 'argv', ['check-changelog', __file__])
    cc.main()


def test_main_no_such_file(monkeypatch, capsys):
    monkeypatch.setattr(sys, 'argv', ['check-changelog', '/no/such/file'])
    cc.main()
    out, err = capsys.readouterr()
    assert err == (
        "check-changelog: [Errno 2] No such file or directory:"
        " '/no/such/file'\n"
    )


def test_main_no_args(monkeypatch, capsys):
    monkeypatch.setattr(sys, 'argv', ['check-changelog'])
    cc.main()


tests/test_machine_summary.py
# Tests for machine-summary
# Run with nosetests
from machine_summary import *


def test_fmt_with_units():
    assert fmt_with_units(12.5, 'megs') == '12.5 megs'
    assert fmt_with_units(12.75, 'megs') == '12.8 megs'
    assert fmt_with_units(12.0, 'megs') == '12 megs'


def test_fmt_size_decimal():
    assert fmt_size_decimal(0) == '0 B'
    assert fmt_size_decimal(1) == '1 B'
    assert fmt_size_decimal(10) == '10 B'
    assert fmt_size_decimal(1000) == '1 KB'
    assert fmt_size_decimal(1200) == '1.2 KB'
    assert fmt_size_decimal(1200*10**3) == '1.2 MB'
    assert fmt_size_decimal(1200*10**6) == '1.2 GB'
    assert fmt_size_decimal(1200*10**9) == '1.2 TB'
    assert fmt_size_decimal(1200*10**12) == '1.2 PB'


def test_fmt_size_si():
    assert fmt_size_si(0) == '0 B'
    assert fmt_size_si(1) == '1 B'
    assert fmt_size_si(10) == '10 B'
    assert fmt_size_si(1024) == '1 KiB'
    assert fmt_size_si(1200) == '1.2 KiB'
    assert fmt_size_si(1200*1024) == '1.2 MiB'
    assert fmt_size_si(1200*1024**2) == '1.2 GiB'
    assert fmt_size_si(1200*1024**3) == '1.2 TiB'
    assert fmt_size_si(1200*1024**4) == '1.2 PiB'


def test_round_binary():
    # these are not very realistic corner cases
    assert round_binary(0) == 0
    assert round_binary(1) == 1
    assert round_binary(2) == 2
    # test cases from actual servers (MemTotal from /proc/meminfo)
    assert round_binary(61680 * 1024) == 64 * 1024**2
    assert round_binary(2061288 * 1024) == 2 * 1024**3
    assert round_binary(3096712 * 1024) == 3 * 1024**3
    assert round_binary(4040084 * 1024) == 4 * 1024**3
    assert round_binary(8061912 * 1024) == 8 * 1024**3



check_changelog.py
#!/usr/bin/python3
from __future__ import print_function
import argparse
import re
import sys


__version__ = '1.0.1'


date_rx = re.compile(r'^\d{4}-\d{2}-\d{2}( \d{2}:\d{2})?')


def warn(filename, lineno, message):
    print("%s:%d: %s" % (filename, lineno, message), file=sys.stderr)


def check(fileobj, filename):
    prev_date = None
    prev_dates = {}
    for n, line in enumerate(fileobj, 1):
        m = date_rx.match(line)
        if not m:
            continue
        cur_date = m.group(0)
        if prev_date is not None and cur_date < prev_date:
            warn(filename, n,
                 'date out of order (%s after %s)' % (cur_date, prev_date))
            if cur_date in prev_dates:
                warn(filename, prev_dates[cur_date],
                     '%s previously used here' % cur_date)
        prev_dates[cur_date] = n
        prev_date = cur_date


def main():
    parser = argparse.ArgumentParser(
        description="Check /root/Changelog for consistency")
    parser.add_argument("--version", action="version", version=__version__)
    parser.add_argument("filename", nargs='*', default=['/root/Changelog'],
                        help="file names to check (default: /root/Changelog)")
    args = parser.parse_args()
    for filename in args.filename or ['/root/Changelog']:
        try:
            with open(filename) as fp:
                check(fp, filename)
        except IOError as e:
            print("%s: %s" % (sys.argv[0], e), file=sys.stderr)


if __name__ == '__main__':
    main()


machine_summary.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
"""
Produce a machine summary table in ReStructuredText
"""

import optparse
import os
import socket
from collections import Counter


__version__ = '0.7.0'


def fmt_with_units(size, units):
    return ('%.1f %s' % (size, units)).replace('.0 ', ' ')


def fmt_size_decimal(bytes):
    size, units = bytes, 'B'
    for prefix in 'KB', 'MB', 'GB', 'TB', 'PB':
        if size >= 1000:
            size, units = size / 1000., prefix
    return fmt_with_units(size, units)


def fmt_size_si(bytes):
    size, units = bytes, 'B'
    for prefix in 'KiB', 'MiB', 'GiB', 'TiB', 'PiB':
        if size >= 1024:
            size, units = size / 1024., prefix
    return fmt_with_units(size, units)


def round_binary(bytes):
    bits = bin(bytes).lstrip('0b')
    if len(bits) < 2:
        return bytes
    mantissa = int(bits[:2], 2)
    exponent = len(bits) - 2
    if bits[2:].strip('0'):
        mantissa += 1
    return mantissa << exponent


def get_hostname():
    """Return the (full) hostname"""
    hostname = socket.getfqdn()
    if hostname in ('localhost6.localdomain6', 'localhost.localdomain'):
        # I'd like to return the full hostname, but socket.getfqdn() is buggy
        # and returns 'localhost6.localdomain6' on some ubuntu versions
        hostname = socket.gethostname()
    return hostname


def read_file(filename):
    with open(filename) as f:
        return f.read()


def get_fields(filename, fieldnames, separator=':'):
    if isinstance(fieldnames, str):
        fieldnames = (fieldnames, )
    with open(filename) as f:
        for line in f:
            field, _, value = line.partition(separator)
            if field.strip() in fieldnames:
                yield value.strip()


def get_cpu_info():
    """Return the model of the first"""
    models = Counter(get_fields('/proc/cpuinfo', ['model name', 'cpu model']))
    return ', '.join(
        '%d × %s' % (count, model)
        for model, count in sorted(models.items())
    )


def get_ram_info():
    """Return the amount of RAM"""
    for value in get_fields('/proc/meminfo', 'MemTotal'):
        kibibytes = int(value.split()[0])
        # How do I add the "reserved" ram that the kernel subtracts from
        # MemTotal?  I see it in /var/log/dmesg, but not in /proc
        # I could use dmidecode or lshw, if I had root...
        # oh well, assume if it's almost a round number of gigs, then it's
        # that exact number of gigs.
        return fmt_size_si(round_binary(kibibytes * 1024))
    return 'n/a'


def get_disk_info(device):
    if device in ('simfs', 'vzfs', '???'):
        # iv.lt VPSes have no /sys/block; they mount /dev/simfs on /
        # hostex.lt VPSes have no /sys/block either; they mount /dev/vzfs on /
        r = os.statvfs('/')
        size = fmt_size_si(r.f_blocks * r.f_bsize)
        return size
    elif device == 'swap':
        with open('/proc/swaps') as f:
            next(f) # skip header line
            size = 0
            for line in f:
                size += int(line.split()[2])
            return fmt_size_si(size * 1024)

    sectors = int(read_file('/sys/block/%s/size' % device))
    size = fmt_size_decimal(sectors * 512)
    if device.startswith('cciss'):
        vendor = read_file('/sys/block/%s/device/vendor' % device).strip()
        raid_level = read_file('/sys/block/%s/device/raid_level' % device).strip()
        return '%s (%s hardware %s)' % (size, vendor, raid_level)
    try:
        model = read_file('/sys/block/%s/device/model' % device).strip()
        return '%s (model %s)' % (size, model)
    except IOError:
        return size


def enumerate_disks():
    # Some containers don't have a /sys/block
    # Some containers have an empty /sys/block
    has_sys_block = os.path.exists('/sys/block')
    if has_sys_block:
        devices = sorted(name for name in os.listdir('/sys/block')
                         if name.startswith(('sd', 'cciss', 'xvd', 'vd')))
    else:
        devices = []
    if os.path.exists('/dev/simfs'):  # OpenVZ
        # Uh, just guessing about swap actually: iv.lt VPSes mount /dev/null on swap
        devices += ['simfs', 'swap']
    if os.path.exists('/dev/vzfs'):   # Xen
        devices += ['vzfs']
    if not devices and not has_sys_block:
        devices += ['???']
    return devices


def get_disks_info():
    disks = enumerate_disks()
    return ',\n        '.join(
        '%s - %s' % (d, get_disk_info(d))
        for d in disks
    )


def get_netdev_info(device):
    mac = read_file('/sys/class/net/%s/address' % device).strip()
    if mac:
        return 'MAC: %s' % mac


def is_interesting_netdev(name):
    return not name.startswith(('lo', 'virbr', 'vboxnet', 'vnet', 'docker'))


def is_bridge(name):
    return name.startswith('br')


def get_network_info():
    devices = sorted(
        (d, get_netdev_info(d)) for d in os.listdir('/sys/class/net')
        if '.' not in d and is_interesting_netdev(d) and not is_bridge(d)
    )
    return ',\n        '.join(
        '%s - %s' % (d, info) if info else d
        for d, info in devices
    )


def get_ip_addresses():
    addresses = []
    for line in os.popen('ip addr'):
        line = line.strip()
        if line.startswith(('inet ', 'inet6 ')) and 'scope global' in line:
            d = line.split()[-1]
            if is_interesting_netdev(d):
                addresses.append((line.split()[1].partition('/')[0], d))
    return addresses


def get_os_info():
    """Return the OS name and version"""
    # bit of a hack that works on ubuntu and openwrt
    if os.path.exists('/etc/lsb-release'):
        fn = '/etc/lsb-release'
    elif os.path.exists('/etc/openwrt_release'):
        fn = '/etc/openwrt_release'
    else:
        return 'n/a'
    for value in get_fields(fn, 'DISTRIB_DESCRIPTION', '='):
        return value.strip('"')
    return 'n/a'


def get_architecture():
    """Return the OS architecture"""
    return os.uname()[4] # (kernel_name, node_name, kernel_release, kernel_version, cpu)


def main():
    parser = optparse.OptionParser('usage: %prog [options]',
        description="Report machine summary information as ReStructuredText")
    parser.add_option('-n', '--no-title', action='store_false', dest='title',
                      default=True, help='skip the title heading')
    opts, args = parser.parse_args()
    if os.getenv('RUN_AS_CGI'):
        print("Content-Type: text/plain; charset=UTF-8")
        print("")
    if opts.title:
        hostname = get_hostname()
        print(hostname)
        print('=' * len(hostname))
        print("")
    print(':CPU: %s' % get_cpu_info())
    print(':RAM: %s' % get_ram_info())
    print(':Disks: %s' % get_disks_info())
    print(':Network: %s' % get_network_info())
    for ipaddr, dev in get_ip_addresses():
        print(':IP: %s (%s)' % (ipaddr, dev))
    print(':OS: %s (%s)' % (get_os_info(), get_architecture()))


if __name__ == '__main__':
    main()
